File: tools/orchestrator/src/schemas/retrievalCandidates.ts
import { z } from "zod";

export const RetrievalCandidateSchema = z.object({
  id: z.string().min(1),
  path: z.string().min(1),
  kind: z.enum(["experience", "education", "project"]),
  group: z.string().min(1),
  title: z.string().min(1),
  tags: z.array(z.string()).default([]),
  matchedFilters: z.array(z.string()).default([])
});

export type RetrievalCandidate = z.infer<typeof RetrievalCandidateSchema>;

export const RetrievalCandidatesSchema = z.object({
  schemaVersion: z.literal(1),
  generatedAt: z.string().min(1),
  query: z.object({
    operation: z.string().min(1),
    variables: z.record(z.string(), z.any()).default({})
  }),
  candidates: z.array(RetrievalCandidateSchema).default([])
});

export type RetrievalCandidates = z.infer<typeof RetrievalCandidatesSchema>;

File: backend/src/resume_orchestrator/schemas/retrieval_candidates.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

Kind = Literal["experience", "education", "project"]


class RetrievalCandidate(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str = Field(min_length=1)
    path: str = Field(min_length=1)
    kind: Kind
    group: str = Field(min_length=1)
    title: str = Field(min_length=1)
    tags: list[str] = Field(default_factory=list)
    matchedFilters: list[str] = Field(default_factory=list)


class RetrievalQuery(BaseModel):
    model_config = ConfigDict(extra="ignore")

    operation: str = Field(min_length=1)
    variables: dict[str, Any] = Field(default_factory=dict)


class RetrievalCandidates(BaseModel):
    model_config = ConfigDict(extra="ignore")

    schemaVersion: Literal[1] = 1
    generatedAt: str = Field(min_length=1)
    query: RetrievalQuery
    candidates: list[RetrievalCandidate] = Field(default_factory=list)
